### PR TITLE
Data entry and dashboard integration

### LINKED ISSUE #89: Integrate Data Entry
Now:
- [ ] Create Data Entry page
- [ ] Add iframe targeting data entry
- [ ] Implement selection of dataset on load

Later:
- [ ] Create Data Entry page
- [ ] Add dropdown on top: OU, periods (only available periods) and teams
- [ ] Fetch forms http://dev2.eyeseetea.com:8082/dhis-web-dataentry/loadForm.action?dataSetId=OssR2l5hHpZ&_=1554712434491
- [ ] Fetch values based on dropdowns selection http://dev2.eyeseetea.com:8082/dhis-web-dataentry/getDataValues.action?periodId=20190406&dataSetId=sK8SUWRPB02&organisationUnitId=nIaj4DxM6tJ&multiOrganisationUnit=false&cc=SS8lww0UED6&cp=CCcHst4fvEk&_=1554712266583
- [ ] Implement push (check if it is included in the form)

### PR DESCRIPTION
**Pull request by adrianq**
### :pushpin: References

* **Issue:** Closes #89

### :memo: Implementation

- Two iframes with a bit of hacking to customise the styling and selection in data entry

TODO: remove non-valid periods and extend test coverage

### :art: Screenshots
Screenshot from 2019-04-12 08-25-26
Screenshot from 2019-04-12 08-25-47

### :fire: Is there anything the reviewer should know to test it?
Install it as a regular DHIS2 app otherwise, you will have problems with the iframe origin policy

### :bookmark_tabs: Others

- Any change in the [metadata repo](https://github.com/EyeSeeTea/vaccination-metadata)? If so, what branch/PR?
- Any change in the [GUI library](https://github.com/EyeSeeTea/d2-ui-components)? If so, what branch/PR?

### CODE REVIEW

**Comment by adrianq** [comment]
@tokland Ready for review!

- I will take care of the periods and help as part of this card https://github.com/EyeSeeTea/vaccination-app/issues/102
- I have created another card to keep in mind we need to restore the tests for the list https://github.com/EyeSeeTea/vaccination-app/issues/101

File: src/components/dashboard/Dashboard.jsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            this.props.snackbar.error(i18n.t("Cannot find dashboard associated to the campaign"));
+        }
+    }
+
+    render() {
+        const { iFrameSrc } = this.state;
+        return (
+            <React.Fragment>
+                <PageHeader
+                    title={i18n.t("Dashboard")}
```
Comment (tokland) [inline src/components/dashboard/Dashboard.jsx]: Makes sense to add the name of the campaign in the title? Dashboard: My test campaign. Same for Data Entry.

File: src/components/data-entry/DataEntry.jsx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import { getOrganisationUnitsById } from "../../models/datasets";
+
+class DataEntry extends React.Component {
+    static propTypes = {
+        d2: PropTypes.object.isRequired,
+        config: PropTypes.object.isRequired,
+    };
+
+    state = {
+        isDataEntryIdValid: true,
```
Comment (tokland) [inline src/components/data-entry/DataEntry.jsx]: Maybe start as false and set to true when loaded?

File: src/components/campaign-configurator/CampaignConfigurator.jsx [hunk 3]
```diff
@@ -73,14 +72,16 @@ class CampaignConfigurator extends React.Component {
         {
             name: "dataEntry",
             icon: "library_books",
+            // TODO: isActive: (d2, dataSet) => canUpdate(d2, d2.models.dataSet, [dataSet]),
```
[the remark this answers is not in the document]
Comment (adrianq) [reply src/components/campaign-configurator/CampaignConfigurator.jsx]: Yep, updated card https://github.com/EyeSeeTea/vaccination-app/issues/102

File: src/models/datasets.js [hunk 4]
```diff
@@ -67,3 +67,23 @@ export function getDashboardId(dataSet, config) {
         .map(av => av.value)
         .first();
 }
+
+export async function getOrganisationUnitsById(id, d2) {
+    try {
+        const fields = ["id,organisationUnits[id,name]"].join(",");
+        const dataSet = await d2.models.dataSets.get(id, { fields });
+        return dataSet.organisationUnits.toArray()[0].id;
+    } catch (err) {
```
[the remark this answers is not in the document]
Comment (adrianq) [reply src/models/datasets.js]: I am doing this to check the id (dataset id or organisation id) is valid and display the snack bar if there is an error here https://github.com/EyeSeeTea/vaccination-app/pull/98/files#diff-5c15787cb2aa9317cc0fd63b4ca8cfd4R26 Better to move the try/catch to the parent function?

File: src/components/data-entry/DataEntry.jsx [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            check();
+        });
+    }
+
+    styleFrame(iframeDocument) {
+        iframeDocument.querySelector(`#header`).remove();
+        iframeDocument.querySelector(`#leftBar`).style.top = "-10px";
+        iframeDocument.querySelector(`body`).style.marginTop = "-55px";
+        iframeDocument.querySelector(`#moduleHeader`).remove();
+    }
```
[the remark this answers is not in the document]
Comment (adrianq) [reply src/components/data-entry/DataEntry.jsx]: Yep, card update: https://github.com/EyeSeeTea/vaccination-app/issues/102

File: src/components/dashboard/Dashboard.jsx (hunk 1 above)
Comment (adrianq) [reply src/components/dashboard/Dashboard.jsx]: It is already present in the iframe... Anyway we can discuss the layout details when we take care of the refinement card.

File: src/models/datasets.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (tokland) [reply src/models/datasets.js]: An explicit snippet could be:

```
const orgUnits = dataSet.organisationUnits.toArray();
return _(orgUnits).isEmpty() ? undefined : orgUnits[0].id;
```

[the remark this answers is not in the document]
Comment (adrianq) [reply src/models/datasets.js]: mmm but the problem is the 'get' not accessing the array
image

[the remark this answers is not in the document]
Comment (tokland) [reply src/models/datasets.js]: Sorry, I was focusing in the getting of the orgUnits. For the non-existing dataset this should work: `const dataSet = await d2.models.dataSets.get(id, { fields }).catch(err => undefined);` And then check if you have dataSet or not in the next line.

[the remark this answers is not in the document]
Comment (adrianq) [reply src/models/datasets.js]: Nice one. Done https://github.com/EyeSeeTea/vaccination-app/pull/98/commits/83ab2176f1648181f9ec046073ada9556ea22a47

**Comment by adrianq** [comment]
All the comments should be resolved now. Btw, I removed the componentDidUpdate and changed it for this approach https://github.com/EyeSeeTea/vaccination-app/pull/98/commits/b405ea336c1b8f04fc111142b2055068d69df641#diff-4f40ff551214939ddd69ed290990568aR22 I think it looks clearer but let me know...

**Comment by tokland** [comment]
>  removed the componentDidUpdate and changed it for this approach 

All good, replacing a DidUpdate by a setState callback is idiomatic.

**Pull request merged**